### PR TITLE
completed frequency RDD and sampling over RDD at large regions

### PR DESCRIPTION
**Pull request by akmorrow13**

### CODE REVIEW

File: mango-core/pom.xml [hunk 1]
```diff
@@ -209,5 +209,10 @@
       <groupId>org.scalatra</groupId>
       <artifactId>scalatra_2.10</artifactId>
     </dependency>
+    <dependency>
```
Comment (akmorrow13) [inline mango-core/pom.xml]: remove, not used

File: mango-core/src/main/scala/org/bdgenomics/mango/layout/AlignmentRecordLayout.scala [hunk 2]
```diff
@@ -111,7 +111,8 @@ object MergedAlignmentRecordLayout extends Logging {
   def apply(rdd: RDD[(ReferenceRegion, CalculatedAlignmentRecord)], binSize: Int): Map[String, List[MutationCount]] = {
 
     // collect and reduce mismatches for each sample
-    val mismatches: RDD[(String, List[MisMatch])] = rdd.map(r => (r._2.record.getRecordGroupSample, r._2.mismatches))
+    val mismatches: RDD[(String, List[MisMatch])] = rdd
```
Comment (akmorrow13) [inline mango-core/src/main/scala/org/bdgenomics/mango/layout/AlignmentRecordLayout.scala]: ?

File: mango-cli/src/main/scala/org/bdgenomics/mango/cli/VizReads.scala [hunk 3]
```diff
@@ -223,6 +199,8 @@ class VizServlet extends ScalatraServlet {
             case Some(_) => {
               val filteredData: RDD[(ReferenceRegion, CalculatedAlignmentRecord)] =
                 AlignmentRecordFilter.filterByRecordQuality(dataOption.get.toRDD(), readQuality)
+                  .filter(_._2.mismatches.size > 0)
+
```
Comment (erictu) [inline mango-cli/src/main/scala/org/bdgenomics/mango/cli/VizReads.scala]: remove extra line

File: mango-cli/src/main/scala/org/bdgenomics/mango/cli/VizReads.scala [hunk 4]
```diff
@@ -270,10 +250,10 @@ class VizServlet extends ScalatraServlet {
               val filteredData: RDD[(ReferenceRegion, CalculatedAlignmentRecord)] =
                 AlignmentRecordFilter.filterByRecordQuality(dataOption.get.toRDD(), readQuality)
               val binSize = VizUtils.getBinSize(region, VizReads.screenSize)
-
               val alignmentData: Map[String, List[MutationCount]] = MergedAlignmentRecordLayout(filteredData, binSize)
 
-              val freqData: Map[String, List[FreqJson]] = FrequencyLayout(filteredData.map(_._2.record), region, binSize, sampleIds)
+              val freqData: Map[String, Iterable[FreqJson]] = VizReads.readsData.getFrequency(region, sampleIds, Option(VizReads.screenSize))
+
```
Comment (erictu) [inline mango-cli/src/main/scala/org/bdgenomics/mango/cli/VizReads.scala]: remove extra line

File: mango-cli/src/main/webapp/WEB-INF/layouts/overall.ssp [hunk 5]
```diff
@@ -7,7 +7,7 @@
 
 <html xmlns="http://www.w3.org/1999/html">
 
-    <% render("../templates/head.ssp")%>
+    <% render("../templates/head.ssp") %>
```
Comment (erictu) [inline mango-cli/src/main/webapp/WEB-INF/layouts/overall.ssp]: remove space

File: mango-cli/src/main/webapp/js/reference.js [hunk 6]
```diff
@@ -101,12 +101,14 @@ function renderLowResRef(data, refContainer, refDiv) {
...
-  .attr("fill", function(d) {
-    if (d.reference === "N") return nColor;
-    else return baseColors[d.reference];
-  });
+    if (d.position < 0) return refWidth;
+    else return Math.max(1, refWidth/(viewRegEnd-viewRegStart));
+    })
+    .attr("fill", function(d) {
+      if (d.reference === "N") return nColor;
+      else if (d.position == -1) return "#47244C";
```
Comment (erictu) [inline mango-cli/src/main/webapp/js/reference.js]: should probably create variable for color

File: mango-core/src/main/scala/org/bdgenomics/mango/RDD/FrequencyRDD.scala [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+   * Gets frequency data for samples over a given region.
+   *
+   * @param region: ReferenceRegion to query over
+   * @param samples: List[String of all samples to query for
+   *
+   * @return Map of (Sample, and list of FreqJson) at every location over the queried region
+   */
+  def get(region: ReferenceRegion, samples: List[String], sampleSize: Option[Int] = None): Map[String, Iterable[FreqJson]] = {
+    var data: RDD[(ReferenceRegion, SampleFreq)] = freqRDD.filterByInterval(region).toRDD
+    data = data.filter(r => samples.contains(r._2.sample))
```
Comment (erictu) [inline mango-core/src/main/scala/org/bdgenomics/mango/RDD/FrequencyRDD.scala]: why not use IntervalRDD.filter? Was this too slow? We should investigate the performance tradeoff.

File: mango-cli/src/main/webapp/WEB-INF/layouts/overall.ssp (hunk 5 above)
Comment (akmorrow13) [inline mango-cli/src/main/webapp/WEB-INF/layouts/overall.ssp]: on purpose

File: mango-core/src/main/scala/org/bdgenomics/mango/RDD/FrequencyRDD.scala (hunk 7 above)
Comment (akmorrow13) [inline mango-core/src/main/scala/org/bdgenomics/mango/RDD/FrequencyRDD.scala]: filters are faster on non-interval specific functions

File: mango-core/src/main/scala/org/bdgenomics/mango/models/AlignmentRecordMaterialization.scala [hunk 8]
```diff
@@ -128,8 +176,21 @@ class AlignmentRecordMaterialization(s: SparkContext, d: SequenceDictionary, par
     seqRecord match {
       case Some(_) => {
         val (reg, ref) = refRDD.getPaddedReference(region)
+
         ks.map(k => {
-          val data = loadFromFile(region, k).map(r => (r._1, CalculatedAlignmentRecord(r._2, MismatchLayout(r._2, ref.get, reg))))
+          val alignmentData = loadFromFile(region, k)
+            .map(r => (ReferenceRegion(r), r))
```
Comment (erictu) [inline mango-core/src/main/scala/org/bdgenomics/mango/models/AlignmentRecordMaterialization.scala]: why the decision to put the map here? Is there actually any way we can avoid the ReferenceRegion construction? Just a thought.

File: mango-core/src/main/scala/org/bdgenomics/mango/RDD/FrequencyRDD.scala (hunk 7 above)
Comment (erictu) [inline mango-core/src/main/scala/org/bdgenomics/mango/RDD/FrequencyRDD.scala]: do you know what the difference is?

**Comment by erictu** [comment]
thanks @akmorrow13 !

**Pull request merged**